authentication/ldap_authentication_provider.py

```diff
@@ -12,35 +12,156 @@
+        if self.search_props:
+            user_dn = self.get_dn_by_user_attribute(who,
+                                                    ldap_url=self.ldap_url,
+                                                    **self.search_props)
+        else:
+            user_dn = who
+        # attempt to bind using the matching user ID/DN and provided password.
+        self.validate_bind(user_dn, password)
+        return who
+
+    @classmethod
+    def handle_authentication_error(cls, err, conn):
+        conn.unbind()
+        cls._logger.warning(err)
+        raise AuthenticationException(err)
+
+    def validate_bind(self, who, password):
+            conn = self.create_bind_conn(self.ldap_url, who, password)
+            conn.unbind()
+
+    @classmethod
+    def get_dn_by_user_attribute(cls,
+                                 who,
+                                 ldap_url,
+                                 base_dn,
+                                 admin_user_id,
+                                 admin_password,
+                                 user_id_attribute=DEFAULT_AUTH_ATTR_NAME):
+
+        search_filter = '{0}={1}'.format(user_id_attribute, who)
+        try:
+            conn = cls.create_bind_conn(ldap_url,
+                                        admin_user_id,
+                                        admin_password)
+            # Search for matches under the base_dn subtree using the attribute
+            # search filter.
+            ldap_result_id = conn.search(base_dn,
+                                         ldap.SCOPE_SUBTREE,
+                                         search_filter,
+                                         [DN_ATTRIBUTE_NAME])
+        except AuthenticationException as e:
+            err = 'Failed binding to \'{0}\' using the dedicated ' \
+                  'admin_user_id: {1}, provided in the configuration; {2}' \
+                .format(ldap_url,
+                        admin_user_id, e)
+            cls._logger.warning(err)
+            raise RuntimeError(err)
+        except ldap.LDAPError as e:
+            err = 'Failed searching for user under base DN: {0}, according ' \
+                  'to the following attribute filter pattern {1}; {2}' \
+                .format(base_dn, search_filter, e)
+            cls.handle_authentication_error(err, conn)
+
+        try:
+            # Retrieve query results according to the result ID in order to
+            # obtain the user DN to bind with.
+            result_type, result_data = conn.result(ldap_result_id, 1)
+            if result_data:
+                if result_type != ldap.RES_SEARCH_RESULT:
+                    raise RuntimeError('An unexpected result object was '
+                                       'returned by the search: {result_type}'
+                                       .format(result_type=result_type))
+            else:
+                raise RuntimeError('An empty result was returned by the search'
```

Comment:
an empty result is a bit of a developer-oriented message. We can rephrase it to something like:
`User \'{user_id}\' not found.format...`